### PR TITLE
Decouple LTC from TS Backend using Lazy IR Builder

### LINKED ISSUE #74628: Decouple LTC from TS backend
### 🚀 The feature, motivation and pitch

Currently, there are a few instances where components from the TorchScript LTC backend are used for infrastructure that should be backend agnostic. For example, [`LazyTensor::CreateTensorNode`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/core/tensor.cpp#L426) relies on `DeviceData`, which is a class based on `TsNode`. This causes some issues when implementing a custom LTC backend, especially when the base class of custom nodes is not `TsNode`.

I'm not sure what the best approach to fix this would be, so I think it would be good to get some other ideas.

CC: @wconstab @antoniojkim 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by antoniojkim**
Next stage of breaking up https://github.com/pytorch/pytorch/pull/74710

IR builder class introduced to decouple the explicit usage of `TsNode` in core lazy tensors.

Requires https://github.com/pytorch/pytorch/pull/75324 to be merged in first.

**Background**
- there are ~ 5 special ops used in lazy core but defined as :public {Backend}Node.  (DeviceData, Expand, Scalar...) 
- we currently require all nodes derive from {Backend}Node, so that backends can make this assumption safely
- it is hard to have shared 'IR classes' in core/ because they depend on 'Node'

**Motivation**

1. avoid copy-paste of "special" node classes for each backend
2. in general decouple and remove all dependencies that LTC has on the TS backend

**Summary of changes**
- new 'IRBuilder' interface that knows how to make 5 special ops
- move 'special' node classes to `ts_backend/`
- implement TSIRBuilder that makes the special TS Nodes
- new backend interface API to get the IRBuilder
- update core code to call the builder

CC: @wconstab @JackCaoG @henrytwo

Partially Fixes #74628

### CODE REVIEW

**Comment by wconstab** [comment]
I think we have 3 possible directions to consider

1) every node is derived from {Backend}Node.  Then we have to fix our core -> MakeNode problem (we need a way for core code to construct backend nodes through an abstraction)

2) some nodes are special ‘core’ nodes. (e.g. devicedata)  Then we have to make sure backends don’t rely on casting every node to {Backend}Node

3) core code can create {Backend}Node as long as it does so in a limited small functions which backends can override (e.g. by provinding {Backend}Tensor class that overrides some methods on LazyTensor class.

File: tools/codegen/api/lazy.py [hunk 1]
```diff
@@ -242,3 +245,71 @@ def keyword_values(self) -> List[LazyArgument]:
...
+
+        self.name = name
+        self.inputs = [f"input{i}" for i in range(self.num_inputs)]
+        self.returns = [f"output{i}" for i in range(self.num_outputs)]
+
+        self.aten_name = op["sym"]
+        if not self.aten_name.startswith("at::"):
+            self.ltc_name = self.aten_name
+
+        args: List[Namespace] = []
```
Comment (wconstab) [inline tools/codegen/api/lazy.py:264]: is Namespace being used in its intended way here? I wonder why this isn't `List[LazyArgument]` like other LazySchemas

File: tools/codegen/dest/lazy_ir.py [hunk 2]
```diff
@@ -355,3 +358,75 @@ def __call__(self, f: NativeFunction) -> List[str]:
...
+            f"""
+            static inline {self.signature} {{
+                static IrBuilder* builder = getBackend()->GetIrBuilder();
+                return builder->Make{self.schema.node_name}({self.args});
+            }}
+            """
+        ).strip()
+
+
+def generate_lazy_ir_node_functions(non_native: Dict[str, Any]) -> str:
```
Comment (wconstab) [inline tools/codegen/dest/lazy_ir.py:400]: I like the style of these templates and thanks for using dedent, i've been meaning to do that for the rest of the codegen!

File: tools/codegen/dest/lazy_ir.py [hunk 3]
```diff
@@ -355,3 +358,75 @@ def __call__(self, f: NativeFunction) -> List[str]:
             return ["\n".join([f"{shape_sig.shape_decl};"])]
         else:
             return []
+
+
+@dataclass
+class LazyIrBuilder:
+    schema: NonNativeLazyIrSchema
```
Comment (wconstab) [inline tools/codegen/dest/lazy_ir.py:365]: I guess i'm wondering (generally) why you needed a new schema type.  i will probably start to realize why when i read more of the code..

File: tools/codegen/gen_lazy_tensor.py [hunk 4]
```diff
@@ -345,17 +358,51 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
...
+                node_base_hdr,
+            ] + op_headers
+            if path
+        ],
+        'ir_builder_declarations': dest.generate_lazy_ir_builder(non_native, lazy_ir_obj),
+        'namespace_prologue': ns_helper.prologue,
+        'namespace_epilogue': ns_helper.epilogue,
+    })
+
+    if backend_name == "TorchScript":
```
Comment (wconstab) [inline tools/codegen/gen_lazy_tensor.py:387]: hmm, probably better to use an explicit option in the codegen for enabling special functionality; but let me first think about why this needs to be special in the first place

File: test/cpp/lazy/test_lazy_ops_util.cpp [hunk 5]
```diff
@@ -1,8 +1,8 @@
 #include <test/cpp/lazy/test_lazy_ops_util.h>
 
 #include <torch/csrc/lazy/backend/lowering_context.h>
-#include <torch/csrc/lazy/ts_backend/ops/device_data.h>
 #include <torch/csrc/lazy/core/ir_dump_util.h>
+#include <torch/csrc/lazy/core/ir_nodes.h>
```
[the remark this answers is not in the document]
Comment (antoniojkim) [reply test/cpp/lazy/test_lazy_ops_util.cpp:5]: as to why we need a new one is because its autogenerated. But you're right. It may also also introduce a circular include.

I think I had a slightly different intention for this autogenerated file when I first made it. `ir_builder.h` is a better name for what it currently does

File: tools/codegen/api/lazy.py (hunk 1 above)
Comment (antoniojkim) [reply tools/codegen/api/lazy.py:264]: > is Namespace being used in its intended way here?

Not exactly. The reason I didn't use `LazyArgument` here is because I thought it would be more difficult to construct one than to just fill in the fields that I needed using an arbitrary object.

I can try to make it a `LazyArgument` if you'd prefer

File: tools/codegen/dest/lazy_ir.py (hunk 3 above)
Comment (antoniojkim) [reply tools/codegen/dest/lazy_ir.py:365]: same reason as above. It would just be more difficult to construct a `LazyIrSchema` given what I had. Which is why I did the above.

However, I might be able to make `NonNativeLazyIrSchema` a subclass of `LazyIrSchema` just to make things a bit nicer.

File: tools/codegen/gen_lazy_tensor.py (hunk 4 above)
Comment (antoniojkim) [reply tools/codegen/gen_lazy_tensor.py:387]: The reason it needs to be special is because the file is being generated to `torch/csrc/lazy/core` which presumably only the TS backend will have access to

Comment (wconstab) [reply tools/codegen/gen_lazy_tensor.py:387]: > The reason it needs to be special is because the file is being generated to `torch/csrc/lazy/core` which presumably only the TS backend will have access to

I would think that all backends have access to stuff in `torch/csrc/lazy/core`, and actually the 'generated' folder is more backend-specific since it gets generated per-backend.  I'm confused about this part.

Comment (antoniojkim) [reply tools/codegen/gen_lazy_tensor.py:387]: Only the TS backend has "write-access" to `torch/csrc/lazy/core`. In my latest local changes, I did change this to a boolean flag to make the intent clearer

File: torch/csrc/lazy/core/ir_builder.h [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  virtual NodePtr MakeNarrow(const Value& input0, const std::vector<int64_t>& base_indices, const std::vector<int64_t>& sizes) const = 0;
+  virtual NodePtr MakePermute(const Value& input0, const std::vector<int64_t>& dims) const = 0;
+  virtual NodePtr MakeResize(const Value& input0, const std::vector<int64_t>& size) const = 0;
+  virtual NodePtr MakeSelectViewUpdate(const Value& input0, const Value& input1, const int64_t& dim, const int64_t& start, const int64_t& end, const int64_t& stride) const = 0;
+  virtual NodePtr MakeSelect(const Value& input0, const int64_t& dim, const int64_t& start, const int64_t& end, const int64_t& stride) const = 0;
+  virtual NodePtr MakeSqueeze(const Value& input0, const int& dim) const = 0;
+  virtual NodePtr MakeUnsqueeze(const Value& input0, const int& dim) const = 0;
+};
+
+static inline NodePtr MakeDeviceData(const std::shared_ptr<BackendData>& data) {
```
Comment (wconstab) [inline torch/csrc/lazy/core/ir_builder.h:35]: i wonder if its that useful to keep these helper functions? What about just having a 'GetIrBuilder' helper in here and then replace all the call sites to GetIrBuilder()->Make{OpName}() ?

File: torch/csrc/lazy/core/ir_builder.h [hunk 7]
```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <c10/core/ScalarType.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/lazy/backend/backend_interface.h>
+#include <torch/csrc/lazy/core/ir.h>
+#include <vector>
+
+// This file contains autogenerated LazyTensor IrBuilder
```
Comment (wconstab) [inline torch/csrc/lazy/core/ir_builder.h:9]: probably worth changing this comment to point out that this file is part of backend interface and ops shouldn't be added without due process, except the view ops which we plan to remove soon pending functionalization

File: torch/csrc/lazy/core/ir_builder.h (hunk 6 above)
Comment (antoniojkim) [reply torch/csrc/lazy/core/ir_builder.h:35]: I primarily included them to keep the `core/` core less verbose. e.g. `MakeExpand` is less verbose than `GetIrBuilder()->MakeExpand`

**Comment by JackCaoG** [comment]
Took a quick pass, lgtm from the high level. Thanks for making this pr 😄 .

**Comment by antoniojkim** [comment]
> Took a quick pass, lgtm from the high level. Thanks for making this pr 😄 .

@JackCaoG can you or someone else on the PyTorch/XLA team check that this doesn't break PyTorch/XLA? There are a lot of calls to `getBackend()` introduced which `torch_xla` will throw an error for

**Comment by JackCaoG** [comment]
`
2022-04-18T22:26:42.1968524Z [1m/var/lib/jenkins/workspace/test/cpp/lazy/test_lazy_ops_util.cpp:5:10: [0m[0;1;31mfatal error: [0m[1m'torch/csrc/lazy/core/ir_nodes.h' file not found[0m
2022-04-18T22:26:42.1968861Z #include <torch/csrc/lazy/core/ir_nodes.h>
2022-04-18T22:26:42.1969166Z [0;1;32m         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

failure seems to be related to some header include issue. I think the fix should be similar to https://github.com/pytorch/pytorch/pull/74756.

**Comment by antoniojkim** [comment]
> ```
> 2022-04-18T22:26:42.1968524Z �[1m/var/lib/jenkins/workspace/test/cpp/lazy/test_lazy_ops_util.cpp:5:10: �[0m�[0;1;31mfatal error: �[0m�[1m'torch/csrc/lazy/core/ir_nodes.h' file not found�[0m
> 2022-04-18T22:26:42.1968861Z #include <torch/csrc/lazy/core/ir_nodes.h>
> 2022-04-18T22:26:42.1969166Z �[0;1;32m         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> failure seems to be related to some header include issue. I think the fix should be similar to #74756.

yeah thats my bad. Didn't catch that locally as my build was using the cached version of that source file. Fixed now

File: torch/csrc/lazy/ts_backend/ts_backend_impl.cpp [hunk 8]
```diff
@@ -99,6 +105,14 @@ class TSBackendImpl : public torch::lazy::BackendImplInterface {
     return std::make_shared<TSData>(scalar, device);
   }
 
+  torch::lazy::BackendDataPtr GetComputationDataFromNode(Node* node) const {
```
Comment (wconstab) [inline torch/csrc/lazy/ts_backend/ts_backend_impl.cpp:108]: @JackCaoG have you seen this new backend API proposed? i think it is a good idea. will you be OK implementing it?

Comment (JackCaoG) [reply torch/csrc/lazy/ts_backend/ts_backend_impl.cpp:108]: yea, I think it is cleaner than trying to dynamiccast in the ltc core, pt/xla can implement it.

File: torch/csrc/lazy/python/init.cpp [hunk 9]
```diff
@@ -10,10 +10,10 @@
 #include <torch/csrc/lazy/backend/backend_interface.h>
 #include <torch/csrc/lazy/core/ir_dump_util.h>
 #include <torch/csrc/lazy/core/internal_ops/ltc_ops.h>
-#include <torch/csrc/lazy/ts_backend/ops/device_data.h>
```
Comment (wconstab) [inline torch/csrc/lazy/python/init.cpp:13]: @antoniojkim This change is the right thing to do, but it also broke internal CI since there is a usage of TSComputation in one of the python bindings below.

The fix is to stop compiling that ts binding in fbcode builds.
https://github.com/pytorch/pytorch/pull/76337

I can either land 76337 and you rebase, or you can just copy the fix into your PR which would probably save time

Comment (wconstab) [reply torch/csrc/lazy/python/init.cpp:13]: cc @shunting314

File: torch/csrc/lazy/ts_backend/ir_builder.h [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+#include <torch/csrc/lazy/ts_backend/view_ops/diagonal.h>
+#include <torch/csrc/lazy/ts_backend/view_ops/narrow_view_update.h>
+#include <torch/csrc/lazy/ts_backend/view_ops/as_strided.h>
+#include <torch/csrc/lazy/ts_backend/view_ops/unsqueeze.h>
+#include <torch/csrc/lazy/ts_backend/view_ops/select.h>
+#include <torch/csrc/lazy/ts_backend/view_ops/view.h>
+#include <torch/csrc/lazy/ts_backend/ops/cast.h>
+#include <torch/csrc/lazy/ts_backend/ops/device_data.h>
+#include <torch/csrc/lazy/ts_backend/ops/generic.h>
+#include <torch/csrc/lazy/ts_backend/ops/batch_norm_ops.h>
```
Comment (Krovatkin) [inline torch/csrc/lazy/ts_backend/ir_builder.h:25]: out of curiosity, is this header really needed?

Comment (antoniojkim) [reply torch/csrc/lazy/ts_backend/ir_builder.h:25]: good catch. Its not strictly needed here. It was only included as this file was originally autogen'd and it just included all of the ops headers. It shouldn't hurt to include it though. I don't think its worth restarting CI over. It can probably be cleaned up at some future time if need be

**Review by wconstab — APPROVED** [review]

File: tools/build_variables.bzl [hunk 11]
```diff
@@ -471,14 +474,13 @@ lazy_tensor_core_sources = [
...
-    "torch/csrc/lazy/ts_backend/dynamic_ir.cpp",
+    "torch/csrc/lazy/ts_backend/tensor_aten_ops.cpp",
     "torch/csrc/lazy/ts_backend/ts_autograd_functions.cpp",
     "torch/csrc/lazy/ts_backend/ts_backend_impl.cpp",
+    "torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp",
     "torch/csrc/lazy/ts_backend/ts_lowering_context.cpp",
     "torch/csrc/lazy/ts_backend/ts_native_functions.cpp",
     "torch/csrc/lazy/ts_backend/ts_node_lowering.cpp",
-    "torch/csrc/lazy/ts_backend/tensor_aten_ops.cpp",
-    "torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp",
```
Comment (wconstab) [inline tools/build_variables.bzl:481]: also just out of curiosity, what is the reason for this change?

File: tools/build_variables.bzl [hunk 12]
```diff
@@ -433,36 +433,39 @@ lazy_tensor_core_sources = [
...
-    "torch/csrc/lazy/core/view_ops/squeeze.cpp",
-    "torch/csrc/lazy/core/view_ops/unsqueeze.cpp",
-    "torch/csrc/lazy/core/view_ops/select_view_update.cpp",
-    "torch/csrc/lazy/core/view_ops/view.cpp",
     # We should better segment the sources, but for now there are actually dependencies
     # from some core files on some of these ts_backend files
     # so we continue to build these parts of ts_backend in all build configs
     "torch/csrc/lazy/ts_backend/config.cpp",
-    "torch/csrc/lazy/ts_backend/ops/arithmetic_ir_ops.cpp",
+    "torch/csrc/lazy/ts_backend/dynamic_ir.cpp",
```
[the remark this answers is not in the document]
Comment (wconstab) [reply tools/build_variables.bzl:449]: > I believe it was because of a failure in one of the mobile builds. I can't see any CI history, but IIRC that was what it was. Though, if it works without that change, then I'm all for reverting it

hmm. well it seems more right to move it back to ts_backend srcs and then re-triage that mobile build i guess.

the specific error is an undefined symbol `torch::lazy::LowerTSBuiltin` when compiling dynamic_ir.cpp without compiling the rest of ts_backend which supplies the lowering stuff.

i'd like to improve the source segmentation too.. i forgot why i left so many of them in the core build in the first place.

[the remark this answers is not in the document]
Comment (wconstab) [reply tools/build_variables.bzl:449]: @antoniojkim actually does your IR builder finally make it possible to move all the ts_backend/ops/* files into the lazy_tensor_ts_sources list?  I think the reason we needed to compile them in core before was becuase we directly included/linked their symbols from core code.  If your PR adds a layer of indirection to core, it should be possible for core to only depend on the interface.

File: torch/csrc/lazy/ts_backend/ir_builder.cpp [hunk 13]
```diff
@@ -0,0 +1,16 @@
+#include <torch/csrc/lazy/backend/backend_interface.h>
+#include <torch/csrc/lazy/ts_backend/ir_builder.h>
+
+// This file contains the TorchScript IrBuilder
+
+namespace torch {
+namespace lazy {
+
+// Get IrBuilder from backend. Use TorchScriptIrBuilder by default
+const IrBuilder* getIrBuilder() {
```
Comment (wconstab) [inline torch/csrc/lazy/ts_backend/ir_builder.cpp:10]: maybe move this file to csrc/lazy/backend,
and just have it do
TORCH_CHECK(hasBackend()) then always use getBackend()->GetIrBuilder()
?

File: torch/csrc/lazy/ts_backend/view_ops/view.cpp [hunk 14]
```diff
@@ -1,4 +1,4 @@
-#include <torch/csrc/lazy/core/view_ops/view.h>
+#include <torch/csrc/lazy/ts_backend/view_ops/view.h>
```
Comment (desertfire) [inline torch/csrc/lazy/ts_backend/view_ops/view.cpp:1]: Not a blocker, but @JackCaoG , have you decided how to handle view_ops for XLA? Leveraging @bdhirsh 's functionalization pass? Just wondering if view_ops should be shared or backend specific.

**Comment by antoniojkim** [comment]
@desertfire The plan for the view ops (at least in LTC and TS backend) is to use the view ops until the functionalization pass is ready. I also have a PR that will be a follow up to this one that will autogenerate those view ops so that making them backend specific is not too cumbersome.
